falcon/request.py

```diff
@@ -1617,6 +1664,85 @@ def _get_wrapped_wsgi_input(self):
 
+    def _getFieldValue(self, field):
+        """Returns input value for imput text fields and FileStream object
+        for input file and None if no file is selected for upload
+
+        Args:
+            field (cgi.FieldStorage or cgi.MiniFieldStorage).
+        """
+        if field.file and 'filename' in field.disposition_options:
+
+            # read first byte of file buffer to make sure if file content exists
+            first_byte = field.file.read(1)
+
+            if first_byte:
+                # NOTE(kgriffs): This is an uploaded file.
+                return FileStream(
+                    file_buffer=field,
+                    max_size=self.options.file_max_upload_size,
+                    first_byte=first_byte
+                )
+            else:
+                # NOTE(kgriffs): No contents found which means no file uploaded.
+                return None
+        else:
+            # NOTE(kgriffs): This is an text field.
+            # DECODE_HTML_CHARS is used to convert html code to utf-8 charcters
+            return DECODE_HTML_CHARS(field.value)
+
+    def _parse_form_urlencoded_or_multipart_form(self):
+        """Parses urlencoded or multipart form data.
+        It sets text fileds in _form_data dict and file buffers in _files dict.
+        """
+
+        if (
+            self.method not in ('GET', 'HEAD') and
+            self.content_type is not None and
+            (
+                'multipart/form-data' in self.content_type or
+                'application/x-www-form-urlencoded' in self.content_type
+            )
+        ):
+            # NOTE(kgriffs): Store input params in _form_data and files in _files
+            self._files = {}
+            self._form_data = {}
+        else:
+            return
+
+        field_storage = cgi.FieldStorage(fp=self.stream, environ=self.env)
+        for fieldname in field_storage.keys():
+            filedata = field_storage[fieldname]
+            if isinstance(filedata, list):
+                # NOTE(kgriffs): fieldname has multiple values
+                list_data = []
+                list_files = []
+                for item in filedata:
+                    value = self._getFieldValue(item)
+                    if isinstance(value, str):
+                        list_data.append(value)
+                    elif value:
+                        list_files.append(value)
+                if list_data:
+                    self._form_data[fieldname] = list_data
+                if list_files:
+                    self._files[fieldname] = list_files
+            else:
+                # NOTE(kgriffs): fieldname has single values
+                value = self._getFieldValue(filedata)
+                if isinstance(value, str):
+                    self._form_data[fieldname] = value
+                elif value:
+                    self._files[fieldname] = value
+
+        for i in list(self._form_data):
```

Comment:
This loop would probably be more efficient if it were to iterate over `self._form_data.items()`, instead.